Generate Top VMR

meisonlikesicecream opened the pull request:
Added script to generate VMR top functions and VMRouter_parameters.h. Script can create top functions for multiple VMRs at the same time, although they will be kept in separate files. I kept the preprocessor stuff in the top files because that allows one to just change three variables to test a different region without having to generate new files  (https://github.com/cms-L1TK/firmware-hls/blob/4ff53b38dd49dd916a9c73d10379c7deada5a517/TrackletAlgorithm/VMRouterTop.h#L22-L25). 

I've tried it for one VMR in each layer/disk with success. Examples of how to run:

`python3 generate_VMR.py`     (creates files for default VMR_L2PHIA with the names VMRouterTop_L2PHIA.h/cc)
`python3 generate_VMR.py --uut VMR_L1PHIE -o`     (creates files for VMR_L1PHIE and overwrites the VMRouterTop.h/cc files)
`python3 generate_VMR.py --uut VMR_L1PHIE VMR_L1PHID`     (creates files for VMR_L1PHIE VMR_L1PHID)

(I made a separate PR for the generate top VMR as I'm going on holiday and didn't want to slow down the process of merging the changes for the other modules. Also the VMR script is quite big and not the prettiest...)

pwittich commented:
I added one line to make it more python2/3 compatible

there are still two outstanding issues:
```
(py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
No config file found, using default configuration
************* Module generate_VMR
W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
E:671,12: print statement used (print-statement)

-----------------------------------
Your code has been rated at 9.62/10
```
The second is fixable by doing
```
from __future__ import print_function
```
and adding parens around all the print calls. 
The first I am not as sure about.

meisonlikesicecream commented:
> I added one line to make it more python2/3 compatible
> 
> there are still two outstanding issues:
> 
> ```
> (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
> No config file found, using default configuration
> ************* Module generate_VMR
> W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
> E:671,12: print statement used (print-statement)
> 
> -----------------------------------
> Your code has been rated at 9.62/10
> ```
> 
> The second is fixable by doing
> 
> ```
> from __future__ import print_function
> ```
> 
> and adding parens around all the print calls.
> The first I am not as sure about.

I don't get any warnings/errors when I run pylint, so I'm not sure how to investigate this. Is it dependent on the pylint version? (I added the print_function import anyway.)

aehart commented:
> > I added one line to make it more python2/3 compatible
> > there are still two outstanding issues:
> > ```
> > (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
> > No config file found, using default configuration
> > ************* Module generate_VMR
> > W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
> > E:671,12: print statement used (print-statement)
> > 
> > -----------------------------------
> > Your code has been rated at 9.62/10
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The second is fixable by doing
> > ```
> > from __future__ import print_function
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and adding parens around all the print calls.
> > The first I am not as sure about.
> 
> I don't get any warnings/errors when I run pylint, so I'm not sure how to investigate this. Is it dependent on the pylint version? (I added the print_function import anyway.)

I also don't get any warnings or errors, so I think it must depend on the pylint version. I'm using the following version, BTW:
```bash
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.9.6 (default, Jun 29 2021, 00:00:00) 
[GCC 11.1.1 20210531 (Red Hat 11.1.1-3)]
```

meisonlikesicecream commented:
> > > I added one line to make it more python2/3 compatible
> > > there are still two outstanding issues:
> > > ```
> > > (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
> > > No config file found, using default configuration
> > > ************* Module generate_VMR
> > > W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
> > > E:671,12: print statement used (print-statement)
> > > 
> > > -----------------------------------
> > > Your code has been rated at 9.62/10
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The second is fixable by doing
> > > ```
> > > from __future__ import print_function
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and adding parens around all the print calls.
> > > The first I am not as sure about.
> > 
> > 
> > I don't get any warnings/errors when I run pylint, so I'm not sure how to investigate this. Is it dependent on the pylint version? (I added the print_function import anyway.)
> 
> I also don't get any warnings or errors, so I think it must depend on the pylint version. I'm using the following version, BTW:
> 
> ```shell
> $ pylint --version
> pylint 2.6.0
> astroid 2.4.2
> Python 3.9.6 (default, Jun 29 2021, 00:00:00) 
> [GCC 11.1.1 20210531 (Red Hat 11.1.1-3)]
> ```

Mine is 
```
pylint 2.9.3
astroid 2.6.2
Python 3.6.8 (default, Nov 16 2020, 16:55:22) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]```

meisonlikesicecream commented:
I added a generate_VMRCM.py as well now. And a --all flag that creates the files needed for a full nonant.

pwittich commented:
> > > > I added one line to make it more python2/3 compatible
> > > > there are still two outstanding issues:
> > > > ```
> > > > (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
> > > > No config file found, using default configuration
> > > > ************* Module generate_VMR
> > > > W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
> > > > E:671,12: print statement used (print-statement)
> > > > 
> > > > -----------------------------------
> > > > Your code has been rated at 9.62/10
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > The second is fixable by doing
> > > > ```
> > > > from __future__ import print_function
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > and adding parens around all the print calls.
> > > > The first I am not as sure about.
> > > 
> > > 
> > > I don't get any warnings/errors when I run pylint, so I'm not sure how to investigate this. Is it dependent on the pylint version? (I added the print_function import anyway.)
> > 
> > 
> > I also don't get any warnings or errors, so I think it must depend on the pylint version. I'm using the following version, BTW:
> > ```shell
> > $ pylint --version
> > pylint 2.6.0
> > astroid 2.4.2
> > Python 3.9.6 (default, Jun 29 2021, 00:00:00) 
> > [GCC 11.1.1 20210531 (Red Hat 11.1.1-3)]
> > ```
> 
> Mine is
> 
> ```
> pylint 2.9.3
> astroid 2.6.2
> Python 3.6.8 (default, Nov 16 2020, 16:55:22) 
> [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
> ```
My calls are from a python2 release. in fact the pylint version appears to be older.  In any event maybe we should just insist on python version 3? this stuff is only required if we want to support python2; maybe that is not needed?

```
(py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py 
No config file found, using default configuration
************* Module generate_VMR
W:332,33: dict.values referenced when not iterating (dict-values-not-iterating)

------------------------------------------------------------------
Your code has been rated at 9.94/10 (previous run: 9.62/10, +0.32)

(py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMRCM.py 
No config file found, using default configuration
************* Module generate_VMRCM
W:241,29: dict.values referenced when not iterating (dict-values-not-iterating)

-----------------------------------
Your code has been rated at 9.93/10

(py2) ➜  emData git:(vmr_generate_top) python --version
Python 2.7.18 :: Anaconda, Inc.
(py2) ➜  emData git:(vmr_generate_top) pylint --version
No config file found, using default configuration
pylint 1.9.2, 
astroid 1.6.5
Python 2.7.18 |Anaconda, Inc.| (default, Apr 23 2020, 17:44:47) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
(py2) ➜  emData git:(vmr_generate_top) 
```

meisonlikesicecream commented:
> > > > > I added one line to make it more python2/3 compatible
> > > > > there are still two outstanding issues:
> > > > > ```
> > > > > (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py
> > > > > No config file found, using default configuration
> > > > > ************* Module generate_VMR
> > > > > W:310,33: dict.values referenced when not iterating (dict-values-not-iterating)
> > > > > E:671,12: print statement used (print-statement)
> > > > > 
> > > > > -----------------------------------
> > > > > Your code has been rated at 9.62/10
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > The second is fixable by doing
> > > > > ```
> > > > > from __future__ import print_function
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > and adding parens around all the print calls.
> > > > > The first I am not as sure about.
> > > > 
> > > > 
> > > > I don't get any warnings/errors when I run pylint, so I'm not sure how to investigate this. Is it dependent on the pylint version? (I added the print_function import anyway.)
> > > 
> > > 
> > > I also don't get any warnings or errors, so I think it must depend on the pylint version. I'm using the following version, BTW:
> > > ```shell
> > > $ pylint --version
> > > pylint 2.6.0
> > > astroid 2.4.2
> > > Python 3.9.6 (default, Jun 29 2021, 00:00:00) 
> > > [GCC 11.1.1 20210531 (Red Hat 11.1.1-3)]
> > > ```
> > 
> > 
> > Mine is
> > ```
> > pylint 2.9.3
> > astroid 2.6.2
> > Python 3.6.8 (default, Nov 16 2020, 16:55:22) 
> > [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
> > ```
> 
> My calls are from a python2 release. in fact the pylint version appears to be older. In any event maybe we should just insist on python version 3? this stuff is only required if we want to support python2; maybe that is not needed?
> 
> ```
> (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMR.py 
> No config file found, using default configuration
> ************* Module generate_VMR
> W:332,33: dict.values referenced when not iterating (dict-values-not-iterating)
> 
> ------------------------------------------------------------------
> Your code has been rated at 9.94/10 (previous run: 9.62/10, +0.32)
> 
> (py2) ➜  emData git:(vmr_generate_top) pylint --py3k generate_VMRCM.py 
> No config file found, using default configuration
> ************* Module generate_VMRCM
> W:241,29: dict.values referenced when not iterating (dict-values-not-iterating)
> 
> -----------------------------------
> Your code has been rated at 9.93/10
> 
> (py2) ➜  emData git:(vmr_generate_top) python --version
> Python 2.7.18 :: Anaconda, Inc.
> (py2) ➜  emData git:(vmr_generate_top) pylint --version
> No config file found, using default configuration
> pylint 1.9.2, 
> astroid 1.6.5
> Python 2.7.18 |Anaconda, Inc.| (default, Apr 23 2020, 17:44:47) 
> [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
> (py2) ➜  emData git:(vmr_generate_top) 
> ```

I think the warning comes from the fact that `mem_copy_dict.values()` is a list in python 2 but not python 3 (https://stackoverflow.com/questions/17695456/why-does-python-3-need-dict-items-to-be-wrapped-with-list). I could wrap it like `list(mem_copy_dict.values())`, but I don't know if we want to do that?

aehart commented:
> I think the warning comes from the fact that `mem_copy_dict.values()` is a list in python 2 but not python 3 (https://stackoverflow.com/questions/17695456/why-does-python-3-need-dict-items-to-be-wrapped-with-list). I could wrap it like `list(mem_copy_dict.values())`, but I don't know if we want to do that?

In this case, `max(mem_copy_dict.values())` works exactly the same whether `values` returns a list (Python 2) or a view (Python 3). So I think we can safetly ignore this warning.

meisonlikesicecream commented:
@pwittich @aehart I just saw on the news slides from yesterday (because I'm stupid and forgot we changed the date/time of the meetings) that you want the top functions in the same file? I put them in separate files because it would be a mess unless I restructure some stuff. Do you want me to change it?

aehart commented:
> @pwittich @aehart I just saw on the news slides from yesterday (because I'm stupid and forgot we changed the date/time of the meetings) that you want the top functions in the same file? I put them in separate files because it would be a mess unless I restructure some stuff. Do you want me to change it?

I think we can work with each top function having a dedicated file. Especially as we go beyond the summer chain, having a single file will probably be easier to deal with, but I think you can wait to do the necessary restructuring until after this chain is on its feet.

aehart commented on emData/download.sh:
```diff
@@ -161,6 +173,8 @@ then
   rm -f LUTs.tar.gz
 fi
 
+./generate_VMR.py -a -w LUTs/wires.dat
```
I added protection against missing include files to the VMRouter_parameters.h that is generated by generate_VMR.py (and similar protections in files for the TC and MC) because some tables are not even produced when running the emulation with the reduced configuration (e.g., no tables for the disk VMRs). So the above line now looks like this:
```c++
  // TE Inner Memory 3
  ap_uint<1> tmpBendTable3n1[bendCutTableSize] =
#if __has_include("../emData/VMR/tables/VMSTE_L1PHID15n1_vmbendcut.tab")
#  include "../emData/VMR/tables/VMSTE_L1PHID15n1_vmbendcut.tab"
#else
  {};
#endif
```

The CI was failing because the CI configuration for the code quality check was still pointing to the old VMRouterTop.cc, which doesn't exist anymore. I just updated it, so it should pass now.

You're right though that we could just generate exactly the VMRs we need for a given configuration. However, I do like the idea of simply replacing the wiring and rerunning download.sh without modification in order to update all the top functions. That being said, we do still have to update this line in generate_VMR.py when switching to a different configuration:
https://github.com/cms-L1TK/firmware-hls/blob/bfffeda52e8bb0107c1fac48cb0c0ac98c10d7c6/emData/generate_VMR.py#L19-L20
Would it be difficult to have the script automatically get a list of all VMRs in the given wiring when using the --all option?

meisonlikesicecream replied:
The --all option already does that? It calls:
https://github.com/cms-L1TK/firmware-hls/blob/vmr_generate_top/emData/generate_VMR.py#L112
The defulat_vmr_list is just a list of random VMRs that I tend to test whenever I debug the VMR :-)

aehart replied:
My bad. The error I saw got scrambled up in my head.

The actual issue is when there are VMRs in the `default_vmr_list` that do not exist in the wiring. For example, when we switch to the reduced config, there are no disk VMRs and that raises an exception that I think is related to nonexistent memories:
```
Traceback (most recent call last):
  File "/home/hart/Research/CMS/tracklets/firmware-hls/emData/./generate_VMR.py", line 707, in <module>
    writeParameterFile(mem_dict)
  File "/home/hart/Research/CMS/tracklets/firmware-hls/emData/./generate_VMR.py", line 308, in writeParameterFile
    "  return " + str(len(mem_dict["IL_"+vmr])) + ";\n"
KeyError: 'IL_VMR_D1PHIA'
```

Obviously, we can just edit the `default_vmr_list` to get rid of the error, but I think when the --all option is used, it should just effectively ignore `default_vmr_list`, and maybe print a warning about VMRs in that list that are not in the given wiring. Would that be easy to do?

meisonlikesicecream replied:
Aaaah yes now I get it, that should be an easy fix.

meisonlikesicecream replied:
Should be done now

aehart reviewed (approved):
Looks good to me. Approved!
